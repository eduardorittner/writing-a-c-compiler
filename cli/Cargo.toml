[package]
name = "cli"
version = "0.1.0"
edition = "2024"

[dependencies]
lex = { path = "../lex" }
parse = { path = "../parse" }
assembly-ast = { path = "../assembly-ast/" }
tracing-subscriber = "0.3.19"
tracing = "0.1.41"
