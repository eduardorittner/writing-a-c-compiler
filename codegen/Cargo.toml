[package]
name = "codegen"
version = "0.1.0"
edition = "2024"

[dependencies]
x86 = { path = "../x86" }
tracing = "0.1.41"
