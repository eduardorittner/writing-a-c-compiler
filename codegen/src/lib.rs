use std::fmt::Write;

pub mod error;
pub use error::*;

use x86::{FnDef, Instruction, Operand, Program, X86};

pub struct Codegen<'input> {
    input: &'input X86,
    output: String,
}

impl<'input> Codegen<'input> {
    pub fn new(input: &'input X86) -> Self {
        Self {
            input,
            output: String::new(),
        }
    }

    pub fn emit(&'input mut self) -> &'input str {
        self.input.nodes.fmt(&mut self.output);
        self.emit_footer();
        &self.output
    }

    pub fn emit_from_input(input: &'input X86) -> CodegenResult<String> {
        let mut codegen = Codegen::new(input);
        codegen.input.nodes.fmt(&mut codegen.output);
        codegen.emit_footer();
        Ok(codegen.output.clone())
    }

    pub fn output(&'input self) -> &'input str {
        &self.output
    }

    fn emit_function(&mut self, name: &str) {
        #[cfg(target_os = "linux")]
        self.output.push_str(name);

        #[cfg(target_os = "macos")]
        self.output.push_str("_" + name);
    }

    fn emit_footer(&mut self) {
        // Disables executable stack to silence linker warning
        #[cfg(target_os = "linux")]
        self.output
            .push_str("\n.section .note.GNU-stack,\"\",@progbits\n");
    }
}

/// Very similar to the `Display` trait, but we can't impl it for `X86` since both are defined in
/// other crates.
trait Format {
    fn fmt(&self, string: &mut String);
}

impl Format for Program {
    fn fmt(&self, string: &mut String) {
        self.main.fmt(string);
    }
}

impl Format for FnDef {
    fn fmt(&self, string: &mut String) {
        let _ = writeln!(string, ".globl {}", &self.name);
        let _ = writeln!(string, "{}:", &self.name);
        self.body.iter().for_each(|instr| instr.fmt(string));
    }
}

impl Format for Instruction {
    fn fmt(&self, string: &mut String) {
        match self {
            Instruction::Mov { src, dst } => {
                let _ = write!(string, "    movl ");
                src.fmt(string);
                let _ = write!(string, ", ");
                dst.fmt(string);
                let _ = writeln!(string);
            }
            Instruction::Ret => {
                let _ = writeln!(string, "    ret");
            }
        }
    }
}

impl Format for Operand {
    fn fmt(&self, string: &mut String) {
        match self {
            Operand::Immediate(value) => {
                let _ = write!(string, "${}", value);
            }
            Operand::Register => {
                let _ = write!(string, "%eax");
            }
        }
    }
}
