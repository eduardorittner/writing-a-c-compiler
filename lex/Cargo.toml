[package]
name = "lex"
version = "0.1.0"
edition = "2024"

[dev-dependencies]
insta = "1.43.1"

[dependencies]
tracing = "0.1.41"

