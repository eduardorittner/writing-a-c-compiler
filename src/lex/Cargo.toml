[package]
name = "lex"
version = "0.1.0"
edition = "2024"

[dependencies]
tracing = "0.1.41"

[dev-dependencies]
insta = "1.43.1"

# Used since insta recommends building it in release mode for faster tests
[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3
