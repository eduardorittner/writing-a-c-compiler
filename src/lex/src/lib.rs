pub mod line;
pub mod token;

pub use token::{Token, TokenType};
use tracing::{Level, span};

use crate::{line::Line, token::TokenSource};

use output::TokenizedOutput;

pub struct Lexer<'src> {
    source: &'src str,
    rest: &'src str,
    offset: usize,
    output: TokenizedOutput<'src>,
}

impl Lexer<'_> {
    pub fn new(source: &str) -> Lexer<'_> {
        Lexer {
            source,
            rest: source,
            offset: 0,
            output: TokenizedOutput::new(source),
        }
    }

    /// Skips whitespace, updating line info on every newline encountered
    fn skip_whitespace(&mut self) -> Option<char> {
        let mut chars = self.rest.chars();
        let mut c;

        loop {
            c = chars.next()?;

            if !c.is_whitespace() {
                break;
            };

            self.offset += c.len_utf8();

            if c == '\n' {
                let line = if let Some(last_line) = self.output.last_line() {
                    Line {
                        start: last_line.end,
                        end: self.offset,
                    }
                } else {
                    Line {
                        start: 0,
                        end: self.offset,
                    }
                };

                self.output.push_line(line);
            }

            self.rest = &self.rest[c.len_utf8()..];
        }

        Some(c)
    }

    /// Consumes an identifier
    fn consume_ident(&mut self) {
        let mut chars = self.rest.chars();

        let c = chars.next().expect("This function should only be called when we still have at least one (alphabetical) char in the input");
        let start = self.offset;

        loop {
            match chars.next() {
                Some(c) if matches!(c, 'A'..='Z' | 'a'..='z' | '1'..='9' | '_') => (),
                Some(c) if c.is_whitespace() => break,
                _ => break,
            };
            self.offset += c.len_utf8();
        }

        self.offset += c.len_utf8();
        self.rest = &self.source[self.offset..];

        let token_source = TokenSource {
            start,
            end: self.offset,
            line: self.output.current_line(),
        };

        self.output
            .push_token(TokenType::Ident, false, token_source);
    }

    fn consume_numeric_constant(&mut self) {
        let mut chars = self.rest.chars();

        let c = chars.next().expect("This function should only be called when we still have at least one (alpha_numerical) char in the input");
        let start = self.offset;

        loop {
            match chars.next() {
                Some(c) if matches!(c, '1'..='9' | '_') => (), // We allow '_' inside numbers
                _ => break,
            };
            self.offset += c.len_utf8();
        }

        self.offset += c.len_utf8();

        self.rest = &self.source[self.offset..];

        let token_source = TokenSource {
            start,
            end: self.offset,
            line: self.output.current_line(),
        };

        self.output
            .push_token(TokenType::Constant, false, token_source);
    }

    pub fn lex(source: &str) -> TokenizedOutput<'_> {
        let mut lexer = Self::new(source);

        let _ = span!(Level::TRACE, "Lexing").entered();

        lexer.run_lexer();
        lexer.output
    }

    fn run_lexer(&mut self) {
        while let Some(c) = self.skip_whitespace() {
            let c_offset = self.offset;
            let c_str = &self.rest[..c.len_utf8()];
            let mut emit_single_char_token = |ttype: TokenType| {
                self.offset += c_str.len();
                self.rest = &self.rest[c_str.len()..];
                self.output.push_token(
                    ttype,
                    false,
                    TokenSource {
                        start: c_offset,
                        end: c_offset + c.len_utf8(),
                        line: self.output.current_line(),
                    },
                );
            };

            match c {
                '(' => emit_single_char_token(TokenType::OpenParen),
                ')' => emit_single_char_token(TokenType::CloseParen),
                '{' => emit_single_char_token(TokenType::OpenBrace),
                '}' => emit_single_char_token(TokenType::CloseBrace),
                ';' => emit_single_char_token(TokenType::Semicolon),
                '/' => emit_single_char_token(TokenType::FrontSlash),
                '\\' => emit_single_char_token(TokenType::BackSlash),
                '-' => emit_single_char_token(TokenType::Hyphen),
                '_' => emit_single_char_token(TokenType::Underscore),
                ':' => emit_single_char_token(TokenType::Colon),
                '*' => emit_single_char_token(TokenType::Asterisk),
                '\'' => emit_single_char_token(TokenType::Quote),
                '"' => emit_single_char_token(TokenType::DoubleQuote),
                ',' => emit_single_char_token(TokenType::Comma),
                c if matches!(c, 'a'..='z' | 'A'..='Z' | '_') => self.consume_ident(),
                c if c.is_ascii_alphanumeric() => self.consume_numeric_constant(),
                '\0' => break,
                _ => panic!("{c:?}"),
            };
        }
        // Emit last line since it doesn't (necessarily) have a '\n'
        self.output.push_line(Line {
            start: self.output.current_line_offset(),
            end: self.offset,
        });
    }
}

mod output {
    use std::fmt::{self, Display};

    use crate::{Token, TokenType, line::Line, token::TokenSource};

    #[derive(Debug)]
    pub struct TokenizedOutput<'src> {
        source: &'src str,
        tokens: Vec<Token>,
        token_sources: Vec<TokenSource>,
        // We don't need a Vec<LineHandle> since they're a simple range [0..lines.len()]
        lines: Vec<Line>,
    }

    impl<'src> TokenizedOutput<'src> {
        pub(crate) fn new(source: &'src str) -> Self {
            TokenizedOutput {
                source,
                tokens: Vec::new(),
                token_sources: Vec::new(),
                lines: Vec::new(),
            }
        }

        pub fn tokens(&self) -> &[Token] {
            &self.tokens
        }

        pub(crate) fn push_token(
            &mut self,
            ttype: TokenType,
            has_error: bool,
            source: TokenSource,
        ) {
            self.tokens.push(Token {
                ttype,
                has_error,
                handle: self.token_sources.len(),
            });
            self.token_sources.push(source);
        }

        pub(crate) fn push_line(&mut self, line: Line) {
            self.lines.push(line);
        }

        pub(crate) fn last_line(&self) -> Option<Line> {
            self.lines.last().copied()
        }

        pub(crate) fn current_line(&self) -> usize {
            self.lines.len()
        }

        pub(crate) fn current_line_offset(&self) -> usize {
            if let Some(line) = self.last_line() {
                line.end
            } else {
                0
            }
        }
    }

    impl<'src> Display for TokenizedOutput<'src> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            for (token, source) in self.tokens.iter().zip(self.token_sources.iter()) {
                write!(
                    f,
                    "{}: [{}] \"{}\"\n",
                    source.line,
                    token.ttype,
                    source.fmt(self.source)
                )?;
            }
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::Lexer;

    macro_rules! snapshot_test (
        ($string:expr) => {
            insta::assert_debug_snapshot!(Lexer::lex($string));
        };
    );

    #[test]
    fn empty_input() {
        snapshot_test!("");
    }

    #[test]
    fn single_char_tokens() {
        snapshot_test!("(){};(");
    }

    #[test]
    fn single_char_tokens_whitespace() {
        snapshot_test!("( ) { } ; (");
    }

    #[test]
    fn single_char_tokens_multiple_lines() {
        snapshot_test!("(\n)\n{\n}\n;\n\n\n");
    }

    #[test]
    fn simple_ident() {
        snapshot_test!("ident");
    }

    #[test]
    fn multiple_simple_ident() {
        snapshot_test!("ident main func int hi");
    }

    #[test]
    fn complex_idents() {
        snapshot_test!(
            "__underscores __more_under_scores_ some1number234 _under1_score_2_with3_numbers5"
        );
    }
}
