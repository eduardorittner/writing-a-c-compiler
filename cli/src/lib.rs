use ast::Tree;
use codegen::Codegen;
use lex::{Lexer, TokenizedOutput};
use parse::Parser;
use std::{
    error::Error,
    io::{Write, stderr, stdout},
    path::{Path, PathBuf},
    process::Command,
};
use tracing::{error, info};
use x86::{X86, lower};

/// Cli arguments
///
/// rustcc <path> --[lex|parse|codegen]
#[derive(Debug, PartialEq)]
pub struct Args {
    pub file: PathBuf,
    pub mode: CompilationMode,
}

impl Default for Args {
    fn default() -> Self {
        Args {
            file: PathBuf::default(),
            mode: CompilationMode::Full,
        }
    }
}

/// Compilation mode
/// Default is Full, which generates the executable
#[derive(Debug, PartialEq, Default)]
pub enum CompilationMode {
    /// Stop after lexing
    Lex,
    /// Stop after parsing
    Parse,
    /// Stop after generating tacky IR
    Tacky,
    /// Stop after codegen, doesn't emit assembly file
    Codegen,
    /// Emits the generated assembly
    NakedAssembly,
    /// Emits the final linked executable
    #[default]
    Full,
}

// TODO add variants
// 1. Missing file path
// 2. Invalid file path
// 3. Invalid flags
#[derive(Debug, PartialEq)]
pub enum CliError {
    NoFileArg,
    NoSuchFile,
}

impl std::fmt::Display for CliError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CliError::NoFileArg => write!(f, "No file was provided"),
            CliError::NoSuchFile => write!(f, "File doesn't exist"),
        }
    }
}

impl Error for CliError {}

pub fn parse_args(mut args: Vec<String>) -> Result<Args, CliError> {
    let mut constructed_args = Args::default();

    // First argument is the executable name
    args.remove(0);

    if args.is_empty() {
        // TODO No file argument
        return Err(CliError::NoFileArg);
    }

    while let Some(arg) = args.pop() {
        match arg.as_str() {
            "--lex" => constructed_args.mode = CompilationMode::Lex,
            "--parse" => constructed_args.mode = CompilationMode::Parse,
            "--tacky" => constructed_args.mode = CompilationMode::Tacky,
            "--codegen" => constructed_args.mode = CompilationMode::Codegen,
            "-S" => constructed_args.mode = CompilationMode::NakedAssembly,
            "--full" => constructed_args.mode = CompilationMode::Full,
            file => {
                let path: PathBuf = file.into();
                constructed_args.file = path;
            }
        };
    }

    if constructed_args.file == PathBuf::default() {
        return Err(CliError::NoFileArg);
    }

    if !constructed_args.file.exists() {
        error!("No such file: {:?}", constructed_args.file);
        eprintln!("{:?}", constructed_args.file);
        // TODO No such file exists
        return Err(CliError::NoSuchFile);
    }

    error!("Parsed args: {constructed_args:?}");

    Ok(constructed_args)
}

pub fn usage_help() {
    println!("Usage:");
    println!("ccompiler [file] [options]");
    println!();
    println!("Options:");
    println!("  --lex: Only runs the lexer");
    println!("  --parse: Only runs the parser");
    println!("  --tacky: Runs up to tacky lowering");
    println!("  --codegen: Runs up to codegen but doesn't emit any file");
    println!("  -S: Emits naked assembly file");
    println!("  --full: Runs the whole pipeline and outputs final executable");
}

pub fn lex<'src>(src: &'src str) -> Result<TokenizedOutput<'src>, Box<dyn Error>> {
    Ok(Lexer::lex(src)?)
}

pub fn parse<'src>(src: &'src str) -> Result<Tree<'src>, Box<dyn Error>> {
    let tokens = lex(src)?;
    let mut parser = Parser::from_tokens(tokens);
    parser.parse();
    Ok(parser.nodes.clone())
}

pub fn tacky<'src>(src: &'src str) -> Result<X86, Box<dyn Error>> {
    let ast = parse(src)?;
    Ok(lower(&ast))
}

pub fn codegen(src: &str) -> Result<X86, Box<dyn Error>> {
    tacky(src)
}

pub fn naked_assembly(src: &str, output_file: &Path) -> Result<(), Box<dyn Error>> {
    let x86 = codegen(src)?;
    let mut codegen = Codegen::new(&x86);
    codegen.emit();
    let assembly = codegen.output().to_string();
    std::fs::write(output_file, assembly)?;
    Ok(())
}

pub fn full(src: &str, input_file: &Path) -> Result<PathBuf, Box<dyn Error>> {
    let mut assembly_file = PathBuf::from(input_file);
    assembly_file.set_extension("s");

    naked_assembly(src, &assembly_file)?;
    let mut executable_file = PathBuf::from(input_file);
    executable_file.set_extension("");

    let mut linker = Command::new("cc");
    linker.arg(&assembly_file).arg("-o").arg(&executable_file);

    info!(
        "Running command: {:?} {:?}",
        linker.get_program(),
        linker.get_args()
    );

    let linker_output = linker.output();

    info!("Deleting intermediate assembly file: {assembly_file:?}");

    let _ = std::fs::remove_file(&assembly_file);

    match linker_output {
        Ok(ok) => {
            info!("Executable generated");
            stdout().write_all(&ok.stdout);
            stderr().write_all(&ok.stderr);
        }
        Err(e) => {
            error!("Got linker error: {e:?}");
            panic!("Linker error: {e:?}");
        }
    };
    Ok(executable_file)
}
