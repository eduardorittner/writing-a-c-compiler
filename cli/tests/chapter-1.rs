mod common;

mod valid {
    use crate::assert_x86;
    use cli::codegen;

    /// Since all these tests are essentially the same program, they just return 0, we can validate
    /// that they end up being represented exactly the same as x86 ast.
    #[test]
    fn same_representation() {
        let mut programs = Vec::new();

        let return_0 = codegen(RETURN_0).unwrap();
        programs.push(codegen(NO_NEWLINES).unwrap());
        programs.push(codegen(NEWLINES).unwrap());
        programs.push(codegen(SPACES).unwrap());
        programs.push(codegen(TABS).unwrap());

        for p in programs {
            assert_eq!(return_0, p);
        }
    }

    const RETURN_0: &str = "int main(void) {
        return 0;
    }";

    #[test]
    fn return_0() {
        assert_x86!(RETURN_0);
    }

    const RETURN_2: &str = "int main(void) {
        return 2;
    }";

    #[test]
    fn return_2() {
        assert_x86!(RETURN_2);
    }

    const NO_NEWLINES: &str = "int main(void){return 0;}";

    #[test]
    fn no_newlines() {
        assert_x86!(NO_NEWLINES);
    }

    const MULTI_DIGITS: &str = "int main(void) {
        return 100;
    }";

    #[test]
    fn multi_digit() {
        assert_x86!(MULTI_DIGITS);
    }

    const NEWLINES: &str = "int
            main
            (
            void
            )
            {
            return
            0
            ;
        }";

    #[test]
    fn newlines() {
        assert_x86!(NEWLINES);
    }

    const SPACES: &str = " int   main    (  void)  {   return  0 ; } ";

    #[test]
    fn spaces() {
        assert_x86!(SPACES);
    }

    const TABS: &str = " int	main	(	void)	{	return	0	;	} ";

    #[test]
    fn tabs() {
        assert_x86!(TABS);
    }
}

mod invalid_lex {
    use crate::lex_err;
    use cli::lex;

    #[test]
    fn at_sign() {
        let src = "int main(void) { return 0@1; }";

        lex_err!(src, "Invalid char '@'");
    }

    #[test]
    fn backslash() {
        let src = "\\";

        lex_err!(src, "Invalid char '\\'");
    }

    #[test]
    fn backtick() {
        let src = "`";

        lex_err!(src, "Invalid char '`'");
    }

    #[test]
    fn invalid_identifier() {
        let src = "int main(void) { return 1foo; }";

        lex_err!(src, "Invalid char in numeric constant 'f'");
    }

    #[test]
    fn invalid_identifier_2() {
        let src = "int main(void) { return @b; }";

        lex_err!(src, "Invalid char '@'");
    }
}

mod invalid_parse {
    use crate::assert_x86;
    use cli::codegen;
}
