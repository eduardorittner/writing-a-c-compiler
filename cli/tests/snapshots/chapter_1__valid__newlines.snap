---
source: cli/tests/chapter-1.rs
expression: assembly_string(NEWLINES).unwrap()
---
".globl main\nmain:\n    movl $0, %eax\n    ret\n\n.section .note.GNU-stack,\"\",@progbits\n"
